### PR TITLE
Census Data Filtering and Pearson Corrleation SnakeMake Rule

### PR DESCRIPTION
**Pull request by anthonyboos559**
Adds functions to filter the census data into groups that share unique combinations of provided metadata categories. Also adds a new SnakeMake rule that implements the Pearson correlation pipeline. The correlation functions operate on the filtered data produced by these data functions.

Aside from the new files:
 `cli.py` was modified to allow the `cli` to be passed `run= False` and expect the `--ckpt_pth` argument.
The `SpeciesDatapipe` was modified to accept `allow_partials` to be passed to the batching datapipe, allowing it to return partial batches as not every group shares the same sample size. The condition on which it checks `allow_partials` was also modified as setting `allow_partials` to True would skip the batch.
The SnakeFile and associated SnakeMake configs were updated to include the correlation rule and support the arguments it required.

### CODE REVIEW

File: workflow/profile/slurm/config.yaml [hunk 1]
```diff
@@ -39,6 +39,11 @@ set-resources:
     mem: 179GB
     gpus_per_node: ""
     cpus_per_task: 1
+  correlations:
+    partition: gpu
+    mem: 179GB
+    gpus_per_node: tesla_v100s:1
```
Comment (jdenhof) [inline workflow/profile/slurm/config.yaml:45]: Is this necessary? We should avoid forcing runs on V100's unless needed (memory/time).

File: src/cmmvae/runners/cli.py [hunk 2]
```diff
@@ -18,7 +18,7 @@ def __init__(self, extra_parser_kwargs: dict = {}, **kwargs):
         Handles loading trainer, model, and data modules from config file,
         while linking common arguments for ease of access.
         """
-        self.is_run = not kwargs.get("run", False)
+        self.is_run = not kwargs.get("run", True)
```
Comment (jdenhof) [inline src/cmmvae/runners/cli.py:21]: Logic doesn't follow LightningCli precedent where run signify's running sub command. When is_run = true no ckpt_path should be added as the fit subcommand already expects this. This change also makes it so by default the ckpt_path will be added which shouldn't be the case when running a built in subcommand. Am I missing something but this feature already exists by specifying run=False?

File: workflow/profile/slurm/config.yaml (hunk 1 above)
Comment (anthonyboos559) [reply workflow/profile/slurm/config.yaml:45]: The correlations rule performs forward-passes with the filtered data and should run on GPU.
The correlation calculations on the model outputs doesn't use the GPU though, so there may be a more optimal solution.

File: src/cmmvae/runners/cli.py (hunk 2 above)
Comment (anthonyboos559) [reply src/cmmvae/runners/cli.py:21]: `self.is_run` isn't passed to the LightningCli `init()` here.
In the old implementation of the cli on main, the adding of `--ckpt_path` to the parser happens as `if not self.is_run`.
In the previous version where `self.is_run = not kwargs.get("run", False)`, `--ckpt_path` would always be skipped as that condition is always `False` in the default case. Setting `run=False` results in no change from the default behavior with this implementation either.
The condition on whether to add `--ckpt_path` to the parser was changed to `if self.is_run`, which will be `True` when run is provided and set to `False`. In the default case where a sub-command is being run, this should be set to `True`, or not provided and default to `True`.
Looking at your dev branch, it seems this issue was already addressed alongside your other changes in a more elegant manner.
Your comments highlight a larger issue where the correlations rule should just be implemented as a subcommand, or not use the cli at all. There is also an issue where the model does not seem to be actually loading form the provided checkpoint right now.

**Pull request merged**